src/lib/reasoners/satml.ml

```diff
@@ -1735,7 +1780,21 @@ module Make (Th : Theory.S) : SAT_ML with type th = Th.t = struct
     if v.level >= 0 then begin
       assert (v.pa.is_true || v.na.is_true);
       pick_branch_lit env
-    end else
+    end else if Options.get_cdcl_tableaux_inst () then
+      (* Do not decide on irrelevant variables (e.g. if we know [P] is true and
+         the only relevant formula involving [Q] is [P \/ Q], deciding on [Q] is
+         unlikely to be helpful -- especially since irrelevant decision won't be
+         propagated to the theory).
+
+         Note that we can only do this if [get_cdcl_tableaux_inst ()] is [true],
+         because otherwise instanciation requires a complete boolean model. *)
```

Comment:
Why do we need a complete boolean model in this case?

Reply:
Turns out this is a more complex question than I expected :) I used this condition because (a) that it is the condition [used for early stopping](https://github.com/OCamlPro/alt-ergo/blob/bc0e045f05068d9e11a4d5c656a52c8c27303d70/src/lib/reasoners/satml.ml#L1787) when the `lazy_cnf` is empty (so it made sense to keep this condition) and (b) not having this condition was crashing [here](https://github.com/OCamlPro/alt-ergo/blob/bc0e045f05068d9e11a4d5c656a52c8c27303d70/src/lib/reasoners/satml_frontend.ml#L634) in a state that was clearly caused by the frontend expecting some atoms to be decided when they were not, hence my description of "we need a complete boolean model".

Turns out the truth is a bit more complicated, and I will try to explain what happens here.

Here are some facts:

 - The `get_cdcl_tableaux_inst ()` is enabled by default in the CDCL-Tableaux solver
 - This flag instructs the instantiation mechanism to use only "relevant" (for some definition of "relevant") atoms when looking for terms during instantiation
 - The definition of "relevant" that is used above can be found [here](https://github.com/OCamlPro/alt-ergo/blob/bc0e045f05068d9e11a4d5c656a52c8c27303d70/src/lib/reasoners/satml_frontend.ml#L698-L702) and ultimately asks SatML for its [instantiation_context](https://github.com/OCamlPro/alt-ergo/blob/bc0e045f05068d9e11a4d5c656a52c8c27303d70/src/lib/reasoners/satml.ml#L2092-L2110). If `get_cdcl_tableaux_th ()` is enabled then we ask the theory; if it is not enabled then we look into the `env.relevants` field, which corresponds to the definition used in this PR.
 - If the `get_cdcl_tableaux_inst ()` flag is disabled, the `conj` field from the SatML *frontend* is used instead (see `atoms_from_sat_branches`) to find the atoms to use for future instantiation rounds.
 - The `relevants` field (and the `lazy_cnf` field) in SatML are updated when backtracking
 - The `conj` field, being maintained by the frontend and not by SatML itself, is *not* updated when backtracking

Given those facts, we can see that if `get_cdcl_tableaux_inst ()` is `false`, the SatML frontend expects that all assertions from previous instantiations, even previous instantiations that are no longer relevant, are enforced by the solver (otherwise we fail [here](https://github.com/OCamlPro/alt-ergo/blob/bc0e045f05068d9e11a4d5c656a52c8c27303d70/src/lib/reasoners/satml_frontend.ml#L634)). Using `get_cdcl_tableaux_inst ()` as a condition here is also coherent with the fact that it is the condition [used for early stopping](https://github.com/OCamlPro/alt-ergo/blob/bc0e045f05068d9e11a4d5c656a52c8c27303d70/src/lib/reasoners/satml.ml#L1787) when the `lazy_cnf` is empty.

For a concrete example: if we assert `(P 3 ∧ Q 3) ∨ R 3` and we have `∀y, P y → y < 0` and `∀y, Q y → y > 0`. The following happens independent of the value of `get_cdcl_tableaux_inst ()`:

 - The SAT explores the `P 3 ∧ Q 3` branch first, and asserts `P 3` and `Q 3`, which are relevant.
 - The quantifiers are triggered, and we instantiate `P 3 → 3 < 0` and `Q 3 → 3 > 0`.
 - `3 < 0` is a contradiction, so we backtrack and learn `¬(P 3)`.
 -  We go into the `R 3` branch. Neither `P 3` nor `Q 3` are relevant anymore.

Now we go to instantiation again, and here the behavior differs.

 - If `get_cdcl_tableaux_inst ()` is `true`, we look only into *relevant* atoms for instantiation. This means we only look into `R 3`, which has no triggers, and we are done[^1].
 - If `get_cdcl_tableaux_inst ()` is `false`, on the other hand, we look into the `conj` field of the SatML frontend. Since it is not backtracked, we find all the previous instantiations there, in particular `P 3 → 3 < 0` and `Q 3 → 3 > 0`. But if we only decide relevant atoms, we don't have a truth value for `Q 3` (since it is not relevant)!

[^1]: Technically I think we move to greedy quantifier instantiation, but nothing happens there either.

Reply:
Note: I think that "instantiation requires a complete boolean model" is a reasonable summary of this explanation for a code comment, given the amount of far-in-the-code implementation details it depends on.

This also asks the question of whether we want to keep the different SatML switches (looking at `--no-cdcl-tableaux-in-theories`, `--no-cdcl-tableaux-in-instantiation` in particular here, but also also thinking about `--no-minimal-bj`); they have been enabled by default for years, have non-trivial costs in terms of code complexity (as evidenced by the above comment), and I don't think anyone actually uses them (Why3 certainly does not, and we ourselves only enable them in the CI tests).

Reply:
Sure, we should test them and if there aren't useful, we may deprecated them in the next release.